chore: Move the setup/teardown of temporary build directories to a process wrapper

frankdavid opened the pull request:
NODE-1459

Previously each build step created its own temporary directory. Cleaning up these directories happened when the build step process exited. However, when the process terminated irregularly (either due to interrupt or an exception), temporary files would not be cleaned up. Our build process produces large temporary files, so not properly removing them leads to the overflow of /tmp.

This PR introduces a process wrapper which creates a temporary directory and ensures that regardless of the output of the invoked command, the temporary directory gets cleaned up. Processes can now access the temporary directory via the `ICOS_TMPDIR` environment variable.

Bownairo commented on toolchains/sysimage/toolchain.bzl:
```diff
@@ -2,6 +2,27 @@
+    wrapper = """
+        tmpdir=$(mktemp -d "/tmp/icosbuildXXXX") &&
+        trap 'sudo rm -rf "$tmpdir"' INT TERM EXIT &&
+        ICOS_TEMP_DIR="$tmpdir" %s "$@" """
```
I wonder if we should use `$TMPDIR` directly. `mkdir`, and python and Rust's `tempfile` all respect this variable, but it does make it harder to find the wrapper. I don't know how much of a concern that is, though.

frankdavid replied:
Absolutely. I ended up passing --tmpdir to mktemp which according to the documentation `uses $TMPDIR if set, else /tmp`.

Bownairo commented on toolchains/sysimage/toolchain.bzl:
```diff
@@ -531,16 +559,23 @@ def _upgrade_image_impl(ctx):
         inputs = [in_boot_partition, in_root_partition, in_version_file],
         outputs = [out],
-        command = "python3 %s -b %s -r %s -v %s -o %s --dflate %s" % (
+        arguments = [
```
Much nicer 👍

Bownairo reviewed (commented):
Great! Excited to not have to worry about this anymore.

---

I think there's only one more use of `atexit` in the repo, also in one of our tools: [bare_metal_deployment/deploy.py](https://github.com/dfinity/ic/blob/b59c0950befdbe9a2347cd43eacf856cf9191610/ic-os/dev-tools/bare_metal_deployment/deploy.py), we should be able to wrap that one too.

frankdavid commented:
Addressed the comments. I also realized that the string wrapper was not a very good approach so I reworked it and made it more reusable. PTAL. Also, perhaps it makes sense to diff to base (and not the previous review) because I reverted quite a few things.

Bownairo commented on toolchains/sysimage/BUILD.bazel:
```diff
@@ -4,7 +4,7 @@ load("@rules_python//python:defs.bzl", "py_binary")
 
 exports_files([
-    "build_base_container_image.py",
+    "build_container_base_image.py",
```
Huh. If the build was fine with an incorrect name, do these need to be exported at all?

frankdavid replied:
Not necessarily, since there is a build rule and others access only the build rule and not the file. I think in general it'd be better to use build rules instead of naked files but I didn't want to change everything. Lmk if you want me to remove the unnecessary exports.

Bownairo replied:
Let's leave it for this PR, we can come through and audit/cleanup later.

Bownairo commented on ic-os/dev-tools/bare_metal_deployment/tools.bzl:
```diff
@@ -20,14 +20,22 @@ def launch_bare_metal(name, image_zst_file):
+        srcs = ["//toolchains/sysimage:proc_wrapper.sh"],
         args = [
+            "python3",
+            "$(location :" + name + "_main)",
```
nit: Will bazel let us use consistent formatting for both `"$(location :" + name + "_main)",` and `":%s_main" % name`?

frankdavid replied:
Bazel doesn't care but it's a good point, I'll use a variable for consistency.
